### DISCUSSION TITLE
Мои мысли по организации файлов (и немного про стиль кодирования)

### DISCUSSION THREAD

**Discussion by alex7850** [godot-engine-ru/shooter-3d#15] [Ideas]
P.S. Совсем забыл о главном сказать. Предлагаю сцены утилитарных геймплейных прототипов держать в отдельной папке в main, а не в гитовых ветках. Проверить несколько наработок не перескакивая с ветки - очень частая необходимость для того же геймдиза. (я даже не уверен можно ли без перезапусков редактора годо сменить ветку гита)

### ГЛАВНЫЕ МЫСЛИ: (просто мои старые наблюдения, не требую внедрять)
1. Придерживаться стандартных годовских рекомендаций по неймингу. Т.е. ПаскальКейсом названия нод и сцен, снейк_кейсом остальных скриптов, файлов.
[Стайл гайд](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html)

2. Служебные папки, где хранятся параметры запуска дебага для иде, или тестовые данные, логи и т.д начинать с точки. Чтобы все они не попали в билд, занести в игнор лист годо при экспорте.
В гитигнор не заносить .import файлы годо, они нужны для корректного импорта графики в проект.

3. Предлагаю до какой-то степени отказаться от папки scripts как основной свалки, аналогично с assets. В пользу следующих принципов организации

4. Подглядел у одного человека, что можно core скрипты основных классов держать в какой-то другой папке и не отвлекаться на шелуху.

5. Игровые сущности, которые часто требуют помимо самой сцены еще и ассеты и отнаследованнный скрипт с расширением функционала, держать тогда отдельно в своей иерархии подпапок.

Пример игровых сущностей:
![изображение](https://user-images.githubusercontent.com/15798187/151654556-85cea129-54df-408c-ac05-363b458e6bfd.png)
[Из видео достаточно популярного разраба](https://youtu.be/EXgyOrjLV1g)

6. ОПЦИОНАЛЬНО Скрипты объединенные общей задачей стараться объединять в "пакеты" ("пакеты" в питоновском смысле как набор скриптов) и выносить в свою общую подпапку в папке modules, classes или типо того.
 
7. ОПЦИОНАЛЬНО Все модули, которые планируется поддерживать отдельно от проекта, для удобства обновления можно разместить в стандартном годовском addons

НЕ СТОЛЬ ВАЖНОЕ
8. UI сцены почти всегда требуют наличие скрипта - как и сущности держать отдельно.
Причем, какие-то базовые сцены со слайдерами, кнопками для переиспользования в UI засунуть в одну папку, а сами сцены экранов меню - в другую.

(Относительные пути сняли бы много проблем с переносом ассетов из проекта в проект. Предмет обсуждения в годо, но пока еще в 3.0 не добавили.)
**Вывод: Главное неудобство такой организации в размазанных сценах, ассетах по всем папкам проекта. Без дополнительных аддонов "палитры" будут сложности с быстрым доступом к игровым объектам во время левелдизайна**

[Пример аддона](https://godotengine.org/asset-library/asset/1070)
![изображение](https://user-images.githubusercontent.com/15798187/151655409-282d8eb5-041c-42eb-a2da-4af1a3647cc7.png)

**Comment by Pigzinzspace** [comment]
можешь это оформить в коменте как требования без объяснений, и воды
это туда, это туда, сцены верблюдом, остальное через палку, скрипты котрые относятся к сцене в папке со сценой, которые не относятся в общей папке сгруппированные по применению. и так далее
короче сделать из этого инструкцию, чтоб программист в вакуме мог прочитать и понят что куда класть
а в диздок добавлю пункт, **требования по оформлению файлов проекта**    и туда это добавлю, ну и обсудим перед этим тоже

**Reply by alex7850** [reply]
Только если ни у кого нет возражений

**Reply by slapin** [reply]
Возражения найдутся когда ты оформишь тезисно, так пока не все понятно что ты хочешь.
После долгого и нудного кодинга в годо у меня сложилась определенная простая модель проекта когда получается его эффективно наращивать:
* autoloads/ - скрипты/сцены больших подсистем требующих глобального состояния
* characters/ - Сцены персонажей с геймплейными узлами, AnimationTree и пр. наследуются от импортированных сцен
* vehicles/ - Сцены техники геймпоейными узлами, наследуются от импортированных сцен
* scenes/characters - Импортированные сцены персонажей
* scenes/locations - Импортированные части мира
* scenes/objects - Импортированные сцены объектов (всякие штуки обычно небольшого размера)
* scripts - скрипты определяющие отдельные фичи, как правило не имеющие состояния и запускаемые каждый фрейм
* scripts/modules - Скрипты компонентной системы для техники и персонажей.
* world/ - Сцены мира вместе с задающими скриптами этих сцен спаунерми и пр.

Думаю что-то такое надо родить для этого проекта в качестве гайда.

**Reply by Pigzinzspace** [reply]
чем быстрее тем лучше, как только проект начнет заполняться кодом будет уже поздно

**Reply by alex7850** [reply]
А что значит "Импортированные сцены"?

**Reply by slapin** [reply]
Например .gltf2
